cranelift/frontend/src/ssa.rs

```diff
@@ -374,7 +370,10 @@ impl SSABuilder {
     /// No predecessors are declared here and the block is not sealed.
     /// Predecessors have to be added with `declare_block_predecessor`.
     pub fn declare_block(&mut self, block: Block) {
-        self.ssa_blocks[block] = SSABlockData::default();
+        // Ensure the block exists so seal_all_blocks will see it even if no predecessors or
+        // variables get declared for this block. But don't assign anything to it:
+        // SecondaryMap automatically sets all blocks to `default()`.
+        let _ = &mut self.ssa_blocks[block];
```

Comment:
I wonder: is this actually necessary? `SecondaryMap` (or at least, one created with `::with_default()`) will return a readonly borrow of some shared "default" value for an `Index` of a non-existing index, or will dynamically expand as needed for an `IndexMut` access; the lack of prior expansion is unobservable (as opposed to just-in-time expansion), I think. Or is that not the case?

Reply:
I almost had myself convinced that I could delete `declare_block` entirely, because as you say, the allocated size of a `SecondaryMap` is unobservable... almost.

The exception is `SecondaryMap::keys`, which iterates only over indexes that are currently initialized. And `seal_all_blocks` uses `self.ssa_blocks.keys()` to determine which blocks need to be sealed.

So if I don't ensure that the map has grown at least to the declared block's index, then I think there's a sequence of public API calls against a `FunctionBuilder` that would hit a debug assertion. It's something like `create_block`, `switch_to_block`, `ensure_inserted_block`, and `finalize`, without declaring any predecessors or calling `use_var`.

I didn't specifically test that because it was starting to make my head hurt. So maybe it's actually okay, but I didn't want to think about it further.